zipfs-law/bin/plotcounts.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+
+import sys
+import re
+import argparse
+import yaml
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+from scipy.optimize import minimize_scalar
+
+
+def nlog_likelihood(beta, counts):
+    """Log-likelihood function."""
+    
+    likelihood = - np.sum(np.log((1/counts)**(beta - 1) - (1/(counts + 1))**(beta - 1)))
+    
+    return likelihood
+
+
+def get_power_law_params(word_counts):
+    """Get the power law parameters.
+    
+    Moreno-Sanchez et al (2016) define alpha (Eq. 1),
+      beta (Eq. 2) and the maximum likelihood estimation (mle)
+      of beta (Eq. 6).
+      
+    Moreno-Sanchez I, Font-Clos F, Corral A (2016)
+      Large-Scale Analysis of Zipf’s Law in English Texts.
+      PLoS ONE 11(1): e0147073.
+      https://doi.org/10.1371/journal.pone.0147073
+    
+    """
+    
+    mle = minimize_scalar(nlog_likelihood, bracket=(1, 4),
+                          args=(word_counts), method='brent')
+    beta = mle.x
+    alpha = 1 / (beta - 1)
+    
+    return alpha, beta
+
+
+def set_plot_params(param_file):
+    """Set the matplotlib rc parameters."""
+    
+    if param_file:
+        with open(param_file, 'r') as reader:
+            param_dict = yaml.load(reader, Loader=yaml.BaseLoader)
+    else:
+        param_dict = {}
+
+    for param, value in param_dict.items():
+        mpl.rcParams[param] = value 
+            
+            
+def plot_fit(xlim, max_rank, beta):
+    """Plot the power law curve that was fitted to the data.
+    
+    Args:
+      xlim (sequence): x-axis bounds (min, max)
+      max_rank (int): maximum word frequency rank
+      beta (float): estimated beta parameter for the power law
+    
+    """
+
+    xvals = np.arange(xlim[0], xlim[-1])
+    yvals = max_rank * (xvals**(-beta + 1))
+    plt.loglog(xvals, yvals, color='grey')
+    
+
+def main(args):
+    """Run the command line program."""
+
+    set_plot_params(args.rcparams)
+    input_csv = args.infile if args.infile else sys.stdin
+    df = pd.read_csv(input_csv, header=None, names=('word', 'word frequency'))
+    df['rank'] = df['word frequency'].rank(ascending=False)
+    df.plot.scatter(x='word frequency', y='rank', loglog=True,
+                    figsize=[12, 6], grid=True,
+                    xlim=args.xlim, ylim=args.ylim)
+
+    alpha, beta = get_power_law_params(df['word frequency'].values)
+    print('alpha:', alpha)
+    print('beta:', beta)
+    plot_fit(args.xlim, df['rank'].values[-1], beta)
```

Comment:
why `values[-1]` ?

Reply:
Because the rank column has been sorted, values[-1] represents the maximum rank (or the rank of the words that only appear once). We need that when plotting the fitted line.

More specifically, relationships of the form y = bx<sup>m</sup> appear as a straight line on log-log plots. Taking the logarithm of the equation yields log(y) = mlog(x) + log(b). Setting X = log(x) and Y = log(y), which corresponds to using a log–log graph, yields the equation: Y = m X + c, where m is the slope of the line (in our case m = -&beta; + 1) and c = log(b) is the intercept on the (log y)-axis, meaning where log x = 0, so, reversing the logs, b is the y-value corresponding to x = 1 (the maximum rank).